randovania/graph/world_graph.py

```diff
@@ -0,0 +1,451 @@
+from __future__ import annotations
+
+import collections
+import dataclasses
+import typing
+
+from randovania.game_description.db.configurable_node import ConfigurableNode
+from randovania.game_description.db.dock import DockLockType, DockWeakness
+from randovania.game_description.db.dock_node import DockNode
+from randovania.game_description.db.event_node import EventNode
+from randovania.game_description.db.event_pickup import EventPickupNode
+from randovania.game_description.db.hint_node import HintNode
+from randovania.game_description.db.node import NodeContext
+from randovania.game_description.db.node_provider import NodeProvider
+from randovania.game_description.db.pickup_node import PickupNode
+from randovania.game_description.db.resource_node import ResourceNode
+from randovania.game_description.db.teleporter_network_node import TeleporterNetworkNode
+from randovania.game_description.requirements.base import Requirement
+from randovania.game_description.requirements.requirement_and import RequirementAnd
+from randovania.game_description.requirements.requirement_or import RequirementOr
+from randovania.game_description.requirements.resource_requirement import (
+    PositiveResourceRequirement,
+    ResourceRequirement,
+)
+from randovania.game_description.resources.node_resource_info import NodeResourceInfo
+from randovania.game_description.resources.resource_type import ResourceType
+
+if typing.TYPE_CHECKING:
+    from collections.abc import Callable, Iterator, Mapping
+
+    from randovania.game_description.db.area import Area
+    from randovania.game_description.db.node import Node
+    from randovania.game_description.db.node_identifier import NodeIdentifier
+    from randovania.game_description.db.region import Region
+    from randovania.game_description.game_database_view import GameDatabaseView
+    from randovania.game_description.game_patches import GamePatches
+    from randovania.game_description.pickup.pickup_entry import PickupEntry
+    from randovania.game_description.requirements.requirement_set import RequirementSet
+    from randovania.game_description.resources.pickup_index import PickupIndex
+    from randovania.game_description.resources.resource_collection import ResourceCollection
+    from randovania.game_description.resources.resource_info import (
+        ResourceGain,
+        ResourceInfo,
+        ResourceQuantity,
+    )
+
+
+class WorldGraphNodeConnection(typing.NamedTuple):
+    target: WorldGraphNode
+    requirement: Requirement
+    requirement_without_leaving: Requirement
```

Comment:
some quick docstrings here would be nice. at a glance it's not obvious to mii what the difference between `requirement` and `requirement_without_leaving` is